File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/ContextModifiers/BlockContextModifierNode.cs
using System.Collections.Generic;
using SoundPlayground.Core;

namespace SoundPlayground.Parser.AbstractSyntaxTree.ContextModifiers {
    public abstract class BlockContextModifierNode : MusicNode {
        public MusicNode Body { get; set; }

        public abstract void Modify ( Context context );

        public abstract void Restore ( Context context );

        public BlockContextModifierNode ( MusicNode body ) {
            Body = body;
        }

        public override IEnumerable<Note> GetCommands ( Context context )
        {
            Context blockContext = context.Fork();

            Modify( blockContext );

            try {
                if ( Body != null ) {
                    foreach ( var command in Body.GetCommands( blockContext ) ) {
                        yield return command;
                    }
                }
            } finally {
                Restore( blockContext );

                context.Join( blockContext );
            }
        }
    }
}

File: code/SoundPlaygroundPy/parser/abstract_syntax_tree/context_modifiers/block_context_modifier_node.py
from .. import MusicNode
from core import Music

class BlockContextModifierNode( MusicNode ):
    def __init__ ( self, body, position : (int, int) = None ):
        super().__init__( position )

        self.body = body
    
    def modify ( self, context ):
        pass

    def restore ( self, context ):
        pass

    def get_events ( self, context ):
        block_context = context.fork()

        events = self.modify( block_context )

        if isinstance( events, Music ):
            for event in events: yield event

        try:
            if self.body != None:
                events = self.body.eval( block_context )

                if isinstance( events, Music ):
                    for event in events: yield event

            events = self.restore( block_context )

            if isinstance( events, Music ):
                for event in events: yield event
        finally:
            context.join( block_context )
